### PR TITLE
Added support for self serve telemetry + Localization fixes

### PR DESCRIPTION
**Pull request by srinathnarayanan**
This PR does the following
**Localization:**
- Separates out the different localization files by namespace
- Modifes i18n.ts file to ensure transaltions are picked up
- Adds the withTranslation construct and removes the <Translation> component
- Adds dynamic loading of required translation files

**Telemetry:**
- Adds support for sending telemetry messages for a self serve class
- Exposes TelemetryMessageType in the SelfServeTypes
- Adds a separate SelfServetelemetryProcessor and SelfServeTelemetryMessage (that requires the class name info)

PR also changes SelfServeUtil to ensure document.referrer is used to build blade reference urls.

### CODE REVIEW

File: src/SelfServe/SelfServe.tsx [hunk 1]
```diff
@@ -14,14 +16,35 @@ import { SelfServeDescriptor } from "./SelfServeTypes";
...
+  let translations: any;
+  try {
+    translations = await import(`../Localization/${language}/${fileName}`);
+  } catch (e) {
+    translations = await import(`../Localization/en/${fileName}`);
+  }
+  i18n.addResourceBundle(language, namespace, translations.default, true);
+};
+
+const loadTranslationsForClass = async (className: string): Promise<void> => {
```
Comment (swviswan) [inline src/SelfServe/SelfServe.tsx:33]: >loadTranslationsForClass [](start = 6, length = 24)

loadTranslations

Comment (swviswan) [inline src/SelfServe/SelfServe.tsx:33]: >loadTranslationsForClass [](start = 6, length = 24)

loadTranslations

File: src/SelfServe/SelfServe.tsx [hunk 2]
```diff
@@ -14,14 +16,35 @@ import { SelfServeDescriptor } from "./SelfServeTypes";
 import { SelfServeType } from "./SelfServeUtils";
 initializeIcons();
 
+const loadTranslationFile = async (namespace: string): Promise<void> => {
```
Comment (swviswan) [inline src/SelfServe/SelfServe.tsx:19]: >namespace [](start = 35, length = 9)

className

File: src/SelfServe/SelfServeUtils.tsx [hunk 3]
```diff
@@ -197,3 +200,25 @@ export const generateBladeLink = (blade: BladeType): string => {
   const databaseAccountName = userContext.databaseAccount.name;
   return `https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/${subscriptionId}/resourceGroups/${resourceGroupName}/providers/Microsoft.DocumentDb/databaseAccounts/${databaseAccountName}/${blade}`;
 };
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const sendTelemetryMessage = (messageType: TelemetryMessageType, data: any = {}): void => {
+  const action = Action.SelfServe;
+  switch (messageType) {
+    case TelemetryMessageType.Trace:
+      trace(action, ActionModifiers.Mark, data);
```
Comment (swviswan) [inline src/SelfServe/SelfServeUtils.tsx:209]: >ActionModifiers.Mark [](start = 20, length = 20)

remove unnecessary

File: src/SelfServe/SelfServeUtils.tsx [hunk 4]
```diff
@@ -197,3 +200,25 @@ export const generateBladeLink = (blade: BladeType): string => {
   const databaseAccountName = userContext.databaseAccount.name;
   return `https://portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/${subscriptionId}/resourceGroups/${resourceGroupName}/providers/Microsoft.DocumentDb/databaseAccounts/${databaseAccountName}/${blade}`;
 };
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const sendTelemetryMessage = (messageType: TelemetryMessageType, data: any = {}): void => {
```
Comment (swviswan) [inline src/SelfServe/SelfServeUtils.tsx:205]: >any [](start = 78, length = 3)

put a wrapper

SelfServeTelemetryData
{
className: string
payload: any
}

**Review by swviswan — APPROVED** [review]
:shipit:

File: src/SelfServe/SelfServe.tsx [hunk 5]
```diff
@@ -14,14 +16,35 @@ import { SelfServeDescriptor } from "./SelfServeTypes";
...
 initializeIcons();
 
+const loadTranslationFile = async (className: string): Promise<void> => {
+  const language = i18n.languages[0];
+  const fileName = `${className}.json`;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  let translations: any;
+  try {
+    translations = await import(`../Localization/${language}/${fileName}`);
```
Comment (languy) [inline src/SelfServe/SelfServe.tsx:26]: Nice to be able to dynamically import the one we need and not all of them :).

File: src/SelfServe/SelfServe.tsx [hunk 6]
```diff
@@ -14,14 +16,35 @@ import { SelfServeDescriptor } from "./SelfServeTypes";
...
 const getDescriptor = async (selfServeType: SelfServeType): Promise<SelfServeDescriptor> => {
   switch (selfServeType) {
     case SelfServeType.example: {
       const SelfServeExample = await import(/* webpackChunkName: "SelfServeExample" */ "./Example/SelfServeExample");
+      await loadTranslations(SelfServeExample.default.name);
       return new SelfServeExample.default().toSelfServeDescriptor();
     }
     case SelfServeType.sqlx: {
       const SqlX = await import(/* webpackChunkName: "SqlX" */ "./SqlX/SqlX");
+      await loadTranslations(SqlX.default.name);
```
Comment (languy) [inline src/SelfServe/SelfServe.tsx:47]: Is it possible (and would it make sense) to infer the name from the class and load this for all the cases? ie moving `loadTranslations()` out of the switch so we only have one call.

File: src/SelfServe/SelfServeTelemetryProcessor.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
+import { sendMessage } from "../Common/MessageHandler";
+import { configContext } from "../ConfigContext";
+import { SelfServeMessageTypes } from "../Contracts/SelfServeContracts";
+import { appInsights } from "../Shared/appInsights";
+import { Action, ActionModifiers } from "../Shared/Telemetry/TelemetryConstants";
+import { userContext } from "../UserContext";
+import { SelfServeTelemetryMessage } from "./SelfServeTypes";
+
+export function trace(action: Action, data: SelfServeTelemetryMessage): void {
```
Comment (languy) [inline src/SelfServe/SelfServeTelemetryProcessor.ts:9]: nit: Prefer the fat arrow syntax, if possible.

File: src/SelfServe/SelfServeTelemetryProcessor.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      timestamp: timestamp,
+      data: JSON.stringify(decorateData(data)),
+    },
+  });
+
+  appInsights.startTrackEvent(Action[action]);
+  return timestamp;
+}
+
+export function traceSuccess(action: Action, data: SelfServeTelemetryMessage, timestamp?: number): void {
```
Comment (languy) [inline src/SelfServe/SelfServeTelemetryProcessor.ts:39]: These functions all look very similar. There seem to be a 1:1 mapping between `Action` and `ActionModifiers`. Consider reducing the code duplication by defining one single function (if it makes sense). Maybe you can take advantage of the `switch` that calls them. Or use a lookup table to get the right `ActionModifiers` (or use `ActionModifiers` directly if `Action` is redundant).

**Review by languy — APPROVED** [review]

File: src/Localization/en/SelfServeExample.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
+{
+    "North Central US": "North Central US",
+    "West US": "West US",
+    "East US 2": "East US 2",
+    "Current Region": "Current Region",
+    "RegionDropdownInfo": "More regions can be added in the future.",
+    "RegionsAndAccountNameValidationError": "Regions and account name should not be empty.",
+    "DbThroughputValidationError": "Please update throughput for database.",
```
[the remark this answers is not in the document]
Comment (srinathnarayanan) [reply src/Localization/en/SelfServeExample.json:8]: this is only an example for selfserve, for new users of the platform to spin it up and make local changes to see what goes where. So I think this is fine.

File: src/Localization/en/SqlX.json [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    "UpdateFailureMessage": "Dedicated gateway resource updation failed.",
+    "DeleteMessage": "Dedicated gateway resource is being deleted.",
+    "DeleteInitializeTitle": "Deleting resource",
+    "DeleteInitializeMessage": "Dedicated gateway resource will be deleted.",
+    "DeleteSuccessTitle": "Resource deleted",
+    "DeleteSuccesseMessage": "Dedicated gateway resource deleted.",
+    "DeleteFailureTitle": "Failed to delete resource",
+    "DeleteFailureMessage": "Dedicated gateway resource deletion failed.",
+    "CannotSave": "Cannot save the changes to the Dedicated gateway resource at the moment",
+    "DedicatedGatewayEndpoint": "Dedicated gatewayEndpoint",
```
[the remark this answers is not in the document]
Comment (srinathnarayanan) [reply src/Localization/en/SqlX.json:38]: this is a change made by the sqlX team and signed off by their PM. So will keep it as such.

File: src/Localization/en/SqlX.json [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    "NoValue": "",
+    "SKUDetails": "SKU Details: ",
+    "CosmosD4Details": "General Purpose Cosmos Compute with 4 vCPUs, 16 GB Memory",
+    "CosmosD8Details": "General Purpose Cosmos Compute with 8 vCPUs, 32 GB Memory",
+    "CosmosD16Details": "General Purpose Cosmos Compute with 16 vCPUs, 64 GB Memory",
+    "CosmosD32Details": "General Purpose Cosmos Compute with 32 vCPUs, 128 GB Memory",
+    "Cost": "Cost",
+    "CostText": "Hourly cost of the dedicated gateway resource depends on the SKU selection, number of instances per region, and number of regions.",
+    "ConnectionString": "Connection String",
+    "ConnectionStringText": "To use the dedicated gateway, use the connection string shown in ",
```
[the remark this answers is not in the document]
Comment (srinathnarayanan) [reply src/Localization/en/SqlX.json:48]: thats a great point. All these changes were made by the sqlx team, Will pass on the feedback. thanks!

File: src/SelfServe/SelfServe.tsx (hunk 6 above)
Comment (srinathnarayanan) [reply src/SelfServe/SelfServe.tsx:47]: To me , it seems like more lines of code. We need to keep track of both the name and descriptor in each case and finally call loadTranslations() and return the descriptor.

**Pull request merged**